api/v1_projects.inc

```diff
@@ -271,6 +272,94 @@ function render_project_page_json($row)
     ];
 }
 
+// POST v1/projects/:projectid/rounds/:roundid
+// checkout a new page to proofread
+function api_v1_project_round($method, $data, $query_params)
+{
+    global $pguser;
+
+    $project = $data[":projectid"];
+    $round = $data[":roundid"];
+    check_project_user($project, $round->id);
+
+    try {
+        $lpage = get_available_page($project->projectid, $project->state, $pguser, $err);
+    } catch (Exception $exception) {
+        throw new BadRequest($exception->getMessage());
+    }
+
+    if (is_null($lpage)) {
+        throw new BadRequest($err);
+    }
+    return ["pagename" => $lpage->imagefile, "pagestate" => $lpage->page_state];
+}
+
+// PUT projects/:projectid/pages/:pagename/rounds/:roundid
+// the PUTs all can change the page text or page state.
+// they are idempotent in the sense that if the identical request is issued
+// again, either the the state and text will not be changed or the request will
+// be rejected because the page state has changed so there will be no change
+// to the server. Logging side-effects excluded.
+function api_v1_project_page_round($method, $data, $query_params)
+{
+    global $pguser;
+
+    $project = $data[":projectid"];
+    $round = $data[":roundid"];
+    check_project_user($project, $round->id);
+
+    try {
+        $lpage = get_indicated_LPage($project->projectid, $project->state, $data[":pagename"], $query_params["pagestate"], 0);
+
+        // can throw InvalidArgumentException
+        $action = get_enumerated_param($query_params, "action", null, ["return", "save_temp", "save_done", "open", "revert"]);
```

Comment:
In general, `PUT`s should take their inputs from their content not the query params. The input for this should probably look like:
```json
{
  "action": "return",
  "page_text": "this is page text"
}
```

Reply:
I realized why "return" and "open" feel weird to me here: they aren't describing an attribute of a page, they're an action. We already know the action which is "persist new information about the page", we just need to know what new information we need to persist.

We should consider inputs for this function like:
```json
{
  "state": "new state",
  "text": "this is page text"
}
```

This function should take and persist the state and text for the page. Based on the current state of the page and the desired state, it might be returning it (and ignoring the page text), saving it (temp or done), or opening it (current state is done and the request is to save it as temp).

Reply:
Yes, that seems like a good way to do it.

Reply:
There is a difficulty with this: if you're opening the page it can go into state 'temp' or stay in 'out' so difficult to use the new state to specify what we want to do. There is very little value in distinguishing 'out' and 'temp' so we could abandon it. (The only effect from the users point of view is that if it is 'out' the 'revert to last save' button is disabled).

Reply:
To abandon 'out' we could copy the text from previous round when checking out the page. Then, before the user saves the page as temp,  revert to last save would do the same as revert to original.

Reply:
I thought that `out` was the page state where nothing had been saved. I.e. if a user checks out the page and then returns it to the round before any interim saves have been done. `temp` is after a save in progress has been done (which can be done manually, or be triggered by WC or changing PI orientation).

The distinction is that currently, if a proofer abandons a page before doing any kind of save, and it's the only page they do in a project, that project does not appear in their "My Projects" list. If an interim save is done, it does. If we abandon `out`, there will be no way to make that distinction. I suppose an argument could be made that, regardless of whether or not a proofer had done an interim save, if they return the page to the round, it shouldn't be counted in their My Projects list. I don't have a strong opinion one way or the other on that.

Reply:
Thanks, there is more significance than I was aware of. I think the difficulty I was envisaging could easily be overcome by just specify a "out_or_temp" state we want to go into. This text does not represent the actual page_state text but is just a convenience for fitting us into the restful jacket.

Reply:
> ... currently, if a proofer abandons a page before doing any kind of save, and it's the only page they do in a project, that project does not appear in their "My Projects" list. If an interim save is done, it does.

That's correct.

> If we abandon `out`, there will be no way to make that distinction.

No, "My Projects" makes that distinction without looking at page states. (If it had to look at page states, it would probably be fairly slow.)

When you interact with a project in certain ways, the code creates or updates an entry in the `user_project_info table`. In particular, if you save (in-progress or as-done) a page in that project, the `t_latest_page_event` is updated. If you visit a project but never save a page, `t_latest_page_event` will be the default 0. And generally, `my_projects.php` only considers projects for which `t_latest_page_event > 0`, so it ignores projects where you haven't done a save.

Reply:
> Instead of code that keeps changing, can you please present a design we can review and comment on and come to a consensus before throwing code at it?

That would be an excellent plan. How can we do this? I'm not sure if the swagger editor is a good way.
I think the profreading functions are difficult to fit into the noun-based restful model because they have previously been verb-based.
But treating a page-checked-out-to-a-user as a resource which can be created, put into different states and deleted would be a way to go.
The addressing mechanism of the url would identify the resource if it exists.
If we used a url like project/page/round/page_state then the validators would determine if the project exists, if the page exists in the project, if the project was in the specified round and if the page is in the specified state. This would duplicate some of the functions of LPage but that doesn't matter much. (Checks on the user also needed).
We want to keep using LPage because at least one of the things it does (getting an available page) is so complicated that it would be risky to do otherwise.
There are several other things to comment on but we first need to identify the best forum.